src/java/picard/util/JavascriptFilter.java

```diff
@@ -0,0 +1,215 @@
+ * THE SOFTWARE.
+ */
+package picard.util;
+
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.List;
+
+import javax.script.Bindings;
+import javax.script.Compilable;
+import javax.script.CompiledScript;
+import javax.script.ScriptEngine;
+import javax.script.ScriptEngineManager;
+import javax.script.ScriptException;
+import javax.script.SimpleBindings;
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMRecord;
+import htsjdk.samtools.filter.SamRecordFilter;
+import htsjdk.samtools.util.CloserUtil;
+import htsjdk.samtools.util.CollectionUtil;
+import htsjdk.samtools.util.IOUtil;
+import htsjdk.variant.variantcontext.VariantContext;
+import htsjdk.variant.vcf.VCFFilterHeaderLine;
+import htsjdk.variant.vcf.VCFHeader;
+import picard.vcf.filter.VariantFilter;
+
+/**
+ * Javascript filter with HEADER type containing TYPE records. contains two
+ * static method to get a SAM Read filter or a VariantFilter
+ *
+ */
+public abstract class JavascriptFilter<HEADER, TYPE> {
+	public static final String HEADER_KEY = "header";
+	/** compiled user script */
+	private CompiledScript script = null;
+
+	/** javascript engine */
+	private ScriptEngine engine = null;
+
+	/** javascript bindings */
+	protected Bindings bindings;
+
+	/** put header in the bindings */
+	protected JavascriptFilter(final HEADER h) {
+		this.bindings = new SimpleBindings();
+		this.bindings.put(HEADER_KEY, h);
+	}
+
+	/** returns key used for header binding */
+	public String getHeaderKey() {
+		return HEADER_KEY;
+	}
+
+	/** returns key used for record binding */
+	public abstract String getRecordKey();
+
+	/** Evaluates this predicate on the given argument. */
+	protected boolean accept(final TYPE record) {
+		try {
+			this.bindings.put(getRecordKey(), record);
+			Object result = this.script.eval(bindings);
+			if (result == null)
+				return false;
+			;
+			if (result instanceof Boolean) {
+				if (Boolean.FALSE.equals(result))
+					return false;
+			} else if (result instanceof Number) {
+				if (((Number) result).intValue() != 1)
+					return false;
+			} else {
+				return false;
+			}
+			return true;
+		} catch (ScriptException err) {
+			throw new RuntimeException(err);
+		}
+	}
+
+	/** creates the Javascript manager, read and compile the scriptFile */
+	protected void initialize(final File scriptFile) {
+		FileReader scriptReader = null;
+		final ScriptEngineManager manager = new ScriptEngineManager();
+		this.engine = manager.getEngineByName("js");
+		if (this.engine == null) {
+			throw new RuntimeException("The embedded 'javascript' engine is not available in java. "
+			        + "Do you use the SUN/Oracle Java Runtime ?");
+		}
+		if (scriptFile == null)
+			throw new NullPointerException("missing javascript file.");
+		IOUtil.assertFileIsReadable(scriptFile);
+		try {
+			final Compilable compilingEngine = (Compilable) this.engine;
```

Comment:
This cast makes me worried.  Is it the case that all javascript engines are `Compilable`? It looks like Rhino and Nashorn both are.  Are there any others?

Could you also extract this and the above engine getting code into it's own static function `getJavaScriptEngine` that returns a Compileable.